symengine/sets.cpp

```diff
@@ -620,6 +632,101 @@ RCP<const Set> Complement::set_complement(const RCP<const Set> &o) const
     return container_->set_complement(newuniv);
 }
 
+bool has_all_numbers(const set_basic &fcont)
+{
+    for (const auto &elem : fcont) {
+        if (not(is_a_Number(*elem) or is_a<Constant>(*elem))) {
+            return false;
+        }
+    }
+    return true;
+}
+
+ConditionSet::ConditionSet(const RCP<const Basic> sym,
+                           RCP<const Boolean> condition)
+    : sym(sym), condition_(condition)
+{
+    SYMENGINE_ASSIGN_TYPEID()
+    SYMENGINE_ASSERT(ConditionSet::is_canonical(sym, condition))
+}
+
+bool ConditionSet::is_canonical(const RCP<const Basic> sym,
+                                RCP<const Boolean> condition)
+{
+    if (eq(*condition, *boolFalse) or eq(*condition, *boolTrue)
+        or not is_a<Symbol>(*sym)) {
+        return false;
+    } else if (is_a<Contains>(*condition)) {
+        return false;
+    }
+    return true;
+}
+
+hash_t ConditionSet::__hash__() const
+{
+    hash_t seed = CONDITIONSET;
+    hash_combine<Basic>(seed, *sym);
+    hash_combine<Basic>(seed, *condition_);
+    return seed;
+}
+
+bool ConditionSet::__eq__(const Basic &o) const
+{
+    if (is_a<ConditionSet>(o)) {
+        const ConditionSet &other = down_cast<const ConditionSet &>(o);
+        return unified_eq(sym, other.get_symbol())
+               and unified_eq(condition_, other.get_condition());
+    }
+    return false;
+}
+
+int ConditionSet::compare(const Basic &o) const
+{
+    SYMENGINE_ASSERT(is_a<ConditionSet>(o))
+    const ConditionSet &other = down_cast<const ConditionSet &>(o);
+    int c1 = unified_compare(sym, other.get_symbol());
+    if (c1 != 0) {
+        return c1;
+    } else {
+        return unified_compare(condition_, other.get_condition());
+    }
+}
+
+RCP<const Set> ConditionSet::set_union(const RCP<const Set> &o) const
+{
+    return SymEngine::set_union({o, rcp_from_this_cast<const Set>()}, false);
+}
+
+RCP<const Set> ConditionSet::set_intersection(const RCP<const Set> &o) const
+{
+    if (not is_a<ConditionSet>(*o)) {
+        return conditionset(sym, logical_and({condition_, o->contains(sym)}));
+    }
+    throw std::runtime_error("Not implemented Intersection class");
+}
+
+RCP<const Set> ConditionSet::set_complement(const RCP<const Set> &o) const
+{
+    return make_rcp<const Complement>(o, rcp_from_this_cast<const Set>());
+}
+
+RCP<const Boolean> ConditionSet::contains(const RCP<const Basic> &o) const
+{
+    map_basic_basic d;
+    if (is_a<FiniteSet>(*o)) {
+        auto &cont = down_cast<const FiniteSet &>(*o).get_container();
+        for (const auto &c : cont) {
+            d[sym] = c;
+            if (not(eq(*condition_->subs(d), *boolTrue)))
+                return boolean(false);
```

Comment:
Here you are checking the `ConditionSet` contains all of the elements of `o`. This is not what `contains` means. Contains means you check that the element `o` is contained in the set. `o` can be a element only if `sym` is a FiniteSet and this case is already handled in Line 726 and below